### PR TITLE
Update total model parameters computation, adds ability to define model label

### PR DESCRIPTION
**Pull request by JayOrten**
- Before, we were using Torchinfo to obtain the total parameters of the model before we begin training. With the addition of flash attention, it seems that this cause OOM errors on VRAM for very large models. This is likely a problem with Torchinfo. This replaces the Torchinfo code with simple PyTorch code that sums up all model parameters. Functionality remains the same.

- Before, we were creating a unique `model_label` for each run, which includes the time stamp. This is nice, but when the job is requeued through Slurm it creates a new folder and doesn't find the checkpoint, thus restarting the training from the beginning. Instead, the code should reference the same model folder. To facilitate this, I have added a new configuration parameter, `model_label`, which the use can define as the name of the folder to save run info(checkpoints, logs). In my experience, this approach is much easier to use and provides the user with more control. If this isn't supplied, a label will be generated. However, this may begin using an already existing folder, which is the biggest downside. So the user needs to be careful.

### CODE REVIEW

File: src/train_model.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ def train_model(config: Struct):
         model = TransformerModel(config)
     else:
         raise ValueError(f"Model type '{config.model_type}' not supported!")
-
+
```
Comment (KimballNJardine) [inline src/train_model.py]: Remove extra spaces from empty line.

File: src/train_model.py [hunk 2]
```diff
@@ -85,18 +85,18 @@ def train_model(config: Struct):
...
-    model_label = f"{config.model_type}_{total_params}" + \
-        f"_LR{config.learning_rate}_ED{config.embed_dim}" + \
-        f"_FFN{config.ffn_dim}_H{config.heads}_S{config.seq_len}" + \
-        f"_{datetime.now().strftime('%Y-%m-%d-%H:%M:%S')}"
+    if config.model_label is not None:
+        model_label = config.model_label
+    else:
+        model_label = f"{config.model_type}_{total_params}" + \
+            f"_LR{config.learning_rate}_ED{config.embed_dim}" + \
+            f"_FFN{config.ffn_dim}_H{config.heads}_S{config.seq_len}"
```
Comment (KimballNJardine) [inline src/train_model.py:113]: Do we want to remove the time at the end (`datetime.now().strftime('%Y-%m-%d-%H:%M:%S')`)?

**Review by KimballNJardine — CHANGES_REQUESTED** [review]
Minor change requested and question to resolve

File: src/train_model.py (hunk 2 above)
Comment (JayOrten) [reply src/train_model.py:113]: Yes, because that's what it making it unique. It's nice until the requeuing doesn't work, because model_label will have a unique timestamp everytime the script is run. This PR is essentially proposing that the model_label should preferably be defined in the config.

File: src/train_model.py (hunk 1 above)
Comment (JayOrten) [reply src/train_model.py]: done

File: src/train_model.py [hunk 3]
```diff
@@ -86,8 +85,7 @@ def train_model(config: Struct):
     print(tabulate(arg_table, tablefmt="grid"))
 
     # Get number of parameters in model
-    model_parameters = filter(lambda p: p.requires_grad, model.parameters())
-    total_params = sum([np.prod(p.size()) for p in model_parameters])
+    total_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
```
Comment (JayOrten) [inline src/train_model.py:104]: @DrewGalbraith , thanks, looks good

**Review by KimballNJardine — APPROVED** [review]
Looks great. Approved.
